### DISCUSSION TITLE
Ask me anything

### DISCUSSION THREAD

**Discussion by theodorusclarence** [theodorusclarence/theodorusclarence.com#179] [General]
Ask me anything! I'll try my best to answer it

**Comment by jonathancaleb** [comment]
Helo
this is really wonderful work

**Comment by RobJacobson** [comment]
Thank you so much for maintaining this! I'm teaching myself Next/TypeScript/Tailwind, but it's hard to keep all of the pieces clear in my head. This is incredibly helpful! I've been dissecting your boilerplate to learn how everything fits together. 

For my own information, do we need to select anything to enable automatic pruning of css, as is discussed in the article below?

https://blog.saeloun.com/2022/02/03/create-typescript-next.js-tailwind-application.html

**Reply by theodorusclarence** [reply]
Hi, I believe with [Tailwind CSS v3](https://tailwindcss.com/blog/tailwindcss-v3#just-in-time-all-the-time), we don't need to prune CSS because the class that is generated is only the class that we're using.

**Comment by osaxon** [comment]
Just want to say thank you and well done with everything you've achieved so far!
I stumbled across your personal page and github whilst researching Next Js and Tailwind CSS and I feel I've learnt a lot from the things you've shared.
I'm currently developing my own personal travel blog and I've "borrowed" some ideas from your site to use on mine. I hop you don't mind! Specifically, a Like button for blog posts and a dynamic table of contents. 
I'm using react-query now swr though which I would highly recommend trying if you haven't already.  

Here's a snippet of my version of useContentMeta hook using react-query instead of swr. One of the great features is that the mutate function can actually make the call to your api and perform the update.

`import apiClient from "../lib/axios";
import { useQuery, useMutation, useQueryClient } from "@tanstack/react-query";

export default function useContentMeta(slug) {
    const queryClient = useQueryClient();

    // get like count from db
    const { data, status } = useQuery({
        queryKey: ["content_meta", slug],
        queryFn: async () => {
            const response = await apiClient.get(`content/${slug}`);
            return response.data;
        },
        staleTime: 5 * 60 * 1000,
    });

    const {
        data: mutatedData,
        mutate: addLike,
        status: mutateStatus,
    } = useMutation({
        // update the like count in the DB
        mutationFn: () => apiClient.post(`/like/${slug}`),
        // optimistally update the state
        onMutate: async () => {
            // cancel any outgoing queries for fresh data
            await queryClient.cancelQueries({
                queryKey: ["content_meta", slug],
            });
            // snapshot current value
            const previousMeta = queryClient.getQueryData([
                "content_meta",
                slug,
            ]);
            // update optimistally
            if (previousMeta) {
                queryClient.setQueryData(["content_meta", slug], {
                    ...previousMeta,
                    likes: previousMeta.likes + 1,
                });
            }
            return { previousMeta };
        },
    });

    return {
        likes: data?.likes,
        views: data?.views,
        addLike,
        mutateStatus,
    };
}
`

**Reply by theodorusclarence** [reply]
Hi, thanks for the kind words and the snippet!

I'm also using react query for most of my ongoing projects, it's great. However, I haven't had the time to refactor this site 😅. 

Good for you migrating them to react query!

**Comment by sherryuser** [comment]
Hello;) I am currently working with your project and found it incredibly useful for my needs. I've come across a section of the code that uses Cloudinary for image hosting and management. I am looking to modify the project to use local images stored in a directory within the project, rather than hosted images. Could you please provide guidance of how to switch from Cloudinary to local file system for image hosting?

**Reply by theodorusclarence** [reply]
Hey, you can use this component for the image

https://github.com/theodorusclarence/ts-nextjs-tailwind-starter/blob/main/src/components/NextImage.tsx

But it doesn't have the capability to have a blur loader yet. I didn't really do that since I'm using cloudinary.

**Reply by sherryuser** [reply]
If my banner images in public/images/banner/ folder, is it possible to change this link, if yes, how I can do that(with classnames)?

```
const { data: frontmatter } = matter(content);
const bannerLink = `https://res.cloudinary.com/theodorusclarence/image/upload/f_auto,c_fill,ar_4:5,w_1200/theodorusclarence/banner/${frontmatter.banner}`;
const ogLink = `https://og.clarence.link/api/blog?templateTitle=${encodeURIComponent(
  frontmatter.title
)}&banner=${encodeURIComponent(bannerLink)}`;
```

**Reply by theodorusclarence** [reply]
You can use `(domain)/images/banner/…`

Also please don’t use my opengraph deployment for your personal project, you can deploy it yourself by forking this https://github.com/theodorusclarence/og

My opengraph domain (clarence.link) is regularly updated and can cause un-noticed breaking in your project

**Comment by sherryuser** [comment]
can you give me full code if it's possible, and for OG yes, I'm using my own link(my website https://shaxriyor.live and OG is https://og.shaxriyor.live) 

Thank you;)

**Reply by theodorusclarence** [reply]
Nice, just change the bannerLink to

`https://shaxriyor.live/images/banner/banner-name.png`

**Reply by sherryuser** [reply]
![og_image](https://github.com/theodorusclarence/theodorusclarence.com/assets/72977488/76159567-af85-4357-9532-ba6681a305f4)
I got this output image, in original code there is ``f_auto,c_fill,ar_4:5,w_1200`` classnames. I think these classnames changes size of the image, where I can add these classnames?

**Reply by theodorusclarence** [reply]
yeah that’s not possible, I use cloudinary’s feature for that. You need to manually crop your images

**Reply by sherryuser** [reply]
I see.. thank you very much;)

**Comment by MedX101** [comment]
hey theo what would be the cause of views count not increasing when the page is refreshed.  i have to manualy click (allow increase view count) button to turn it off,  but it only  increase only once then goes back to constant state (same views). is it becouse of the same IP adress or doiman name (.vercel.app) ? but on theodorusclarence.com and other website leerob.io it automatically increase when you refresh the page ..   so this case i am referering to theodorusclarence.com.            

additional info: views do appear on database, in case you thought it will be a database issues. coz i think it might be related to local storage logic. so please answer this as if i did not make any changes to theodorusclarence.com. i just wanna know where to fix this .

 thanks

**Reply by theodorusclarence** [reply]
this block of code prevents that  
https://github.com/theodorusclarence/theodorusclarence.com/blob/368f274fdab4c1e2ebaad8c426b8582b10157706/src/pages/_app.tsx#L29-L41

I added that so my vercel previews won’t ever increment the database. 

You can either add your site to the env for whitelist, or change the blockDomainMeta in env.ts to false

**Reply by MedX101** [reply]
thanks theo, it is now working fine.😃

so meaning that only views will increase if the domain name is   " NEXT_PUBLIC_BLOCK_DOMAIN_WHITELIST "  right ?

and if  its set to false : export const blockDomainMeta =
  isProd || process.env.NEXT_PUBLIC_META_BLOCK_DOMAIN === 'false';   meaning views will increase on any domain name right ?

**Reply by theodorusclarence** [reply]
yep

**Comment by sherryuser** [comment]
Hello Theodor, I'm using you repo for may own website, I have one issue left, how I can fix likes and views counter (I mean how and where to setup DB), It'd be great if you can help me with that;)

**Reply by theodorusclarence** [reply]
Hi! I can't give you a full on walkthrough, but every service that you need to setup is listed in the .env.example and the [contributing guide](https://github.com/theodorusclarence/theodorusclarence.com/blob/main/CONTRIBUTING.md)

**Comment by MedX101** [comment]
Hey theo...is there any reason you chose not to included sitemap.xml and robots.txt from pushing them to production.

And how google are indexing your pages, are you submitting each and url of your post page. I noticed you put those (sitemap.xml and robots.txt ) under gitigonre


@theodorusclarence

**Reply by theodorusclarence** [reply]
Hey, yes, I'm using [next-sitemap](https://www.npmjs.com/package/next-sitemap). So it will be automatically generated on build and added the the public folder in Vercel.

**Reply by theodorusclarence** [reply]
I just found out that since I migrated this project using pnpm, it won't correctly generate. I fixed it in this [commit](https://github.com/theodorusclarence/theodorusclarence.com/commit/f412dbff111c2ecab3549ed6b84aa51d6a679582)

**Reply by MedX101** [reply]
So you didn't  do **anything****  on Google Search Console. 

The reason why am asking this because I was trying to submit sitemap on Google Search Console..but I couldn't find it..

**Reply by theodorusclarence** [reply]
Yes, you don't need to do anything if it the setup is correct. You can check my robots.txt, and sitemap

https://theodorusclarence.com/robots.txt
https://theodorusclarence.com/sitemap.xml

If you're forking my project, you might want to replicate my last commit

**Comment by MedX101** [comment]
@theodorusclarence 

How can I customize the url slug.. I have noticed you customized the library into shorts..is there any way I can change url without changing or renaming the the folder which will break the logic..

For example Project to something else..


Just like Nextra Theme, you can just change the url of any page by using Json..

**Reply by theodorusclarence** [reply]
You can’t, you need to change directly in the code.

**Reply by MedX101** [reply]
Okay thank you, I thought it's possible..

**Comment by MedX101** [comment]
Hey @theodorusclarence  quick question here:

Did you made the statistics page data not to update by default? If yes ? How can i change it ? 

because deleted pages data ( views, like ) are still showing there.

**Reply by theodorusclarence** [reply]
Yes, you need to manually delete them from the database

**Reply by theodorusclarence** [reply]
You can email me at me@theodorusclarence.com for further question so it doesn’t seem like a support page here 🤣

**Reply by MedX101** [reply]
Okay I'll use email in future 😀😀..

**Comment by TheJohnBradley** [comment]
@theodorusclarence 

How did you host your new "theodorusclarence.com"? (not the open source one)

**Reply by theodorusclarence** [reply]
Heyy @TheJohnBradley, it’s still hosted on vercel! As for the images I use Cloudinary CDN.

Additionally, both the website and images are cached via cloudflare.

**Reply by TheJohnBradley** [reply]
Thanks a lot!

**Reply by TheJohnBradley** [reply]
@theodorusclarence 

Could you also please share which vercel plan you’re currently using for your site?
Also, do you have an admin dashboard implemented? If so, I’d really appreciate seeing what it looks like—would you mind sharing a few screenshots? 😄

**Reply by theodorusclarence** [reply]
I'm still using the free plan. 

Nope, no dashboard; all content management is done by code.

**Comment by TheJohnBradley** [comment]
@theodorusclarence 

How did you make the signature with the nice animation on your new website (font, maybe code snippet, etc.)

**Reply by theodorusclarence** [reply]
That one is closed source. But i’m pretty sure you can clone it by using inspect element. Good luck!
